apps/api/pages/api/destination-calendars/[id]/_patch.ts

```diff
@@ -57,15 +60,130 @@ import { schemaQueryIdParseInt } from "~/lib/validations/shared/queryIdTransform
  *        description: Destination calendar not found
  */
 export async function patchHandler(req: NextApiRequest) {
-  const { prisma, query, body } = req;
+  const { userId, isAdmin, prisma, query, body } = req;
   const { id } = schemaQueryIdParseInt.parse(query);
   const parsedBody = schemaDestinationCalendarEditBodyParams.parse(body);
+  let credentialId: number | undefined = undefined;
+  const assignedUserId = isAdmin ? parsedBody.userId || userId : userId;
+
+  // when linked with eventTypeId, we need to fetch the userId from the eventTypeId
+  if (parsedBody.eventTypeId) {
+    const eventType = await prisma.eventType.findFirst({
+      where: { id: parsedBody.eventTypeId, userId: assignedUserId },
+    });
+    if (!eventType)
+      throw new HttpError({
+        statusCode: 400,
+        message: "Bad request, eventTypeId invalid",
+      });
+    parsedBody.userId = undefined;
+  }
+
+  const { userCredentials } = await findUserCredentials(req);
+
+  if (parsedBody.integration && !parsedBody.externalId) {
+    throw new HttpError({ statusCode: 400, message: "External Id is required with integration value" });
+  }
+  if (!parsedBody.integration && parsedBody.externalId) {
+    throw new HttpError({ statusCode: 400, message: "Integration value is required with external ID" });
+  }
+
+  if (parsedBody.integration && parsedBody.externalId) {
+    const calendarCredentials = getCalendarCredentials(userCredentials);
+
+    const { connectedCalendars } = await getConnectedCalendars(
+      calendarCredentials,
+      [],
+      parsedBody.externalId
+    );
+    const connectedCalendar = connectedCalendars.find(
+      (c) =>
+        c?.primary?.externalId === parsedBody.externalId && c?.primary?.integration === parsedBody.integration
+    );
+
+    if (!connectedCalendar)
+      throw new HttpError({
+        statusCode: 400,
+        message: "Bad request, credential id invalid",
+      });
+
+    credentialId = connectedCalendar.primary?.credentialId;
+  }
 
   const destinationCalendar = await prisma.destinationCalendar.update({
     where: { id },
-    data: parsedBody,
+    data: { ...parsedBody, credentialId },
```

Comment:
the credentialId here will never be `undefined`, it'll always be picked up internally. In case it is undefined (which means it isn't linked internally), an error will be thrown before we reach this part of the code.